pkg/kubernetes/apply.go

```diff
@@ -2,35 +2,98 @@ package kubernetes
 
 import (
 	"fmt"
-
-	"github.com/fatih/color"
+	"strings"
+	"time"
 
 	"github.com/grafana/tanka/pkg/kubernetes/client"
 	"github.com/grafana/tanka/pkg/kubernetes/manifest"
-	"github.com/grafana/tanka/pkg/term"
 )
 
 // ApplyOpts allow set additional parameters for the apply operation
 type ApplyOpts client.ApplyOpts
 
 // Apply receives a state object generated using `Reconcile()` and may apply it to the target system
 func (k *Kubernetes) Apply(state manifest.List, opts ApplyOpts) error {
-	alert := color.New(color.FgRed, color.Bold).SprintFunc()
-
-	cluster := k.ctl.Info().Kubeconfig.Cluster
-	context := k.ctl.Info().Kubeconfig.Context
-	if !opts.AutoApprove {
-		if err := term.Confirm(
-			fmt.Sprintf(`Applying to namespace '%s' of cluster '%s' at '%s' using context '%s'.`,
-				alert(k.Spec.Namespace),
-				alert(cluster.Name),
-				alert(cluster.Cluster.Server),
-				alert(context.Name),
-			),
-			"yes",
-		); err != nil {
-			return err
+	return k.ctl.Apply(state, client.ApplyOpts(opts))
+}
+
+// Orphaned returns previously created resources that are missing from the
+// local state. It uses UIDs to safely identify objects.
+func (k *Kubernetes) Orphaned(state manifest.List) (manifest.List, error) {
+	apiResources, err := k.ctl.Resources()
+	if err != nil {
+		return nil, err
+	}
+
+	start := time.Now()
+	fmt.Print("fetching UID's .. ")
+	uids, err := k.uids(state)
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println("done", time.Since(start))
+
+	var orphaned manifest.List
+
+	// join all kinds that support LIST into a comma separated string for
+	// kubectl
+	kinds := ""
+	for _, r := range apiResources {
+		if !strings.Contains(r.Verbs, "list") {
+			continue
 		}
+
+		kinds += "," + r.FQN()
 	}
-	return k.ctl.Apply(state, client.ApplyOpts(opts))
+	kinds = strings.TrimPrefix(kinds, ",")
+
+	start = time.Now()
+	fmt.Print("fetching previously created resources .. ")
+	// get all resources matching our label
+	matched, err := k.ctl.GetByLabels("", kinds, map[string]string{
+		LabelEnvironment: k.Env.Metadata.NameLabel(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println("done", time.Since(start))
+
+	// filter unknown
+	for _, m := range matched {
+		// ignore known ones
+		if uids[m.Metadata().UID()] {
+			continue
+		}
+
+		// ComponentStatus resource is broken in Kubernetes versions
+		// below 1.17, it will be returned even if the label does not
+		// match. Ignoring it here is fine, as it is an internal object
+		// type.
+		if m.APIVersion() == "v1" && m.Kind() == "ComponentStatus" {
+			continue
+		} else if m.APIVersion() == "v1" && m.Kind() == "Endpoints" {
```

Comment:
There are use-cases where you might need to create endpoints directly. 
https://kubernetes.io/docs/concepts/services-networking/service/#without-selectors

Though as endpoints dont include any metadata to indicate they are managed by a controller, i am not sure how we would ever know that was the case.

Reply:
True, but `Endpoint` has `ObjectMeta` and thus `metadata.labels` – which we can identify those we did create ourselves – right?

Reply:
Unfortunately, automatically-created Endpoints get all the labels from the original Service.  They are the only core object where this happens (it's probably a bug).

Reply:
Possible approaches:
- just always delete all Endpoints ([qbec does this](https://github.com/splunk/qbec/blob/65ec57bed31ad5bf3fde654695b6b05918b2dff6/internal/remote/client.go#L268))
- use annotations not labels to mark objects ([kubecfg does this](https://github.com/bitnami/kubecfg/blob/9aa6e1279ba939f98b4430b3af099f86ab6011b5/pkg/kubecfg/update.go#L483) although is migrating to labels)
- use both annotation and label (kubecfg will probably do this in future)
- use some other info to determine the original groupkind under which the object was created (kubecfg might also do this - using the serialised info in last-applied-configuration annotation).   This has some impact on migrated objects (eg extensions -> apps) too that should be thought through.
- add the original groupkind into the labels too - so we can ignore objects that have "changed kinds" in the server-side filter (I only just thought of this).  Same caveat as above re migrated apigroups.

Reply:
Because we use `kubectl` for every `apply`, everything manually created should have the `kubectl.kubernetes.io/last-applied-configuration` annotation, right? So I guess it's sufficient to test for its existence?

Reply:
Implemented that, seems to work